### PR TITLE
331-mqtt-to-postgres-ds

### LINKED ISSUE #331: MQTT-to-Postgres: enable digital shadow functionality
**Reason**
We need to store the relevant data from the mqtt broker to the timescaleDB in order to build the digital shadow (issue #297).

**New Topics to be stored in the timescaleDB:**
/productTag (issue #309) -> in productTagTable
/productTagString (issue #319) -> in productTagStringTable
/addParentToChild (issue #320) -> in productInheritanceTable

**Changed Topics to be stored in timescaleDB:**
/uniqueProduct (issue #318) -> in uniqueProductTable, Added the uniqueProductAlternativeID

### PR DESCRIPTION
**Pull request by larswitte**
Fixes #331: Changes MQTT-to-postgres microservice.

### CODE REVIEW

File: golang/.idea/golang.iml [hunk 1]
```diff
@@ -1,5 +1,6 @@
 <?xml version="1.0" encoding="UTF-8"?>
```
Comment (JeremyTheocharis) [inline golang/.idea/golang.iml:1]: Can you delete this .idea folder altogether?

File: golang/cmd/mqtt-to-postgresql/database.go [hunk 2]
```diff
@@ -1909,3 +1999,28 @@ func GetComponentID(assetID int, componentName string) (componentID int) {
 
 	return
 }
+
+// Todo (naming conventions), sachen
+func GetUniqueProductID(aid int, assetID int) (uid int, err error) {
```
Comment (JeremyTheocharis) [inline golang/cmd/mqtt-to-postgresql/database.go:2004]: This function is perfectly suited for caching. When calling it, always check the cache beforehand (you can find the cache in the internal package `golang/PKG/internal/cache.go`)

File: golang/.idea/golang.iml (hunk 1 above)
Comment (larswitte) [reply golang/.idea/golang.iml]: I removed the folder.

File: golang/cmd/mqtt-to-postgresql/database.go (hunk 2 above)
Comment (larswitte) [reply golang/cmd/mqtt-to-postgresql/database.go]: Good idea, thank you! I implemented a uniqueProductID cache (`GetUniqueProductIDFromCache()`, `StoreUniqueProductIDToCache()`).

**Comment by larswitte** [comment]
After this commit b4ea48ba6a346c297f955f861e3e11fa2dcdcc90 `productTag`, `productTagString`, `addParentToChild` and `uniqueProduct` were correctly inserted into the postgresDB (tried out with nodeRed flow simulating two processing steps).
Changes in the db (accessed with DataGrip) were relatively fast (seconds).

**Open points at the commit:**
- Hypertables were disabled
-  the order of the timestamps in the uniqueProductTable were not checked
- the entry `productId` in the uniqueProductTable contained Strings. It should contain a int referencing productID in the productTable

**Comment by larswitte** [comment]
Testing this commit 3773ebb12e2d920ac0b2e2f55e6c0ead70839bf4 with the nodeRed simulation did not function. The productTable should need `addProuct` MQTT Messages, one for each product and asset. After sending them, they didn't appear in the `productTable`. It took more then ten minutes after they appeared in the table. Additionally the asset information (in the `addProduct` Messages was directly used to generate new assets in the asset table (meaning the MQTT messages arrived in time, but were not processed). They also (i think) did not appear in the logs regarding the queue directly. Only after they appeared in the postgresDB I could find them in the logs.

**Comment by larswitte** [comment]
> Testing this commit [3773ebb](https://github.com/united-manufacturing-hub/united-manufacturing-hub/commit/3773ebb12e2d920ac0b2e2f55e6c0ead70839bf4) with the nodeRed simulation did not function (in the commit I tried fixing the in my comment above stated open points). The productTable should need `addProduct` MQTT Messages, one for each product and asset. After sending them, they didn't appear in the `productTable`. It always took more then ten minutes after they appeared in the table. Additionally the asset information (in the `addProduct` Messages was directly used to generate new assets in the asset table (meaning the MQTT messages arrived in time, but were not processed). They also (i think) did not appear in the logs regarding the queue directly. Only after they appeared in the postgresDB I could find them in the logs.
> 
> I went through the go code and did not find significant differences to the processing of other MQTT-Topics. I tried sending the `addProduct` messages one by one and as a batch - no difference. Multiple times restarted the cluster - no difference.
> I also switched to the united-manufacturing-hub main branch - no difference (`addProduct` Messages still did not go into the postgresDB).
> 
> I cant explain that behavior. Obviously I am missing something.
> 
> **Possible failure points:**
> 
>     * Message from node red not correctly sent: unlikely because messages direcly arrive, in the end result in correct entries in the `productTable`
> 
>     * Problem with the queue: seems like it but I could not make out one difference in the code which might be the reason
> 
>     * Problem with sql insert/message processing: unlikely because the time delay problem seemingly already exists in the queue.
> 
> 
> Different Observation: At one time one `id` in the `assetTable` was missing (weird because `id` is the serial primary key. Probably unrelated but I wanted to mention it.

I changed all files on the test server,  deactivated the check of the uniqueProduct Timestamps and split up the nodered addProduct and AssemblyStation Mqtt testmessage generation.

Now it works correctly. I think in the future it might be a good idea to add an automatic functionality: if a `uniqueProduct` Message arrives and no productID can be found for the product name and asset, a new entry in the `productTable` should be automatically created. With this functionality no additional `addProduct` message would be required before new uniqueProduct types.

File: deployment/factorycube-server/templates/mqtttopostgresql-statefulset.yaml [hunk 3]
```diff
@@ -46,16 +46,16 @@ spec:
             topologyKey: "kubernetes.io/hostname"
       containers:
         - name: {{ include "factorycube-server.fullname" . }}-mqtttopostgresql
-          {{- if .Values.mqtttopostgresql.tag }}
-          image: {{ .Values.mqtttopostgresql.image }}:{{ .Values.mqtttopostgresql.tag }}
+          {{- if .Values.mqtttopostgresql.image.tag }}
+          image: {{ .Values.mqtttopostgresql.image.repository }}:{{ .Values.mqtttopostgresql.image.tag }}
           {{- else }}
           image: {{ .Values.mqtttopostgresql.image }}:{{ .Chart.AppVersion }}
```
Comment (JeremyTheocharis) [inline deployment/factorycube-server/templates/mqtttopostgresql-statefulset.yaml:52]: Adjust this line as well 

```yaml
{{ .Values.mqtttopostgresql.image.repository }}
```

File: deployment/factorycube-server/templates/timescale-configmap.yaml [hunk 4]
```diff
@@ -178,7 +178,58 @@ data:
 
     -- #74
     ALTER TABLE counttable ADD IF NOT EXISTS scrap INTEGER DEFAULT 0;
-    
+
+    -- #297
+    ALTER TABLE uniqueProductTable ADD COLUMN IF NOT EXISTS uniqueProductAlternativeID TEXT NOT NULL;
```
Comment (JeremyTheocharis) [inline deployment/factorycube-server/templates/timescale-configmap.yaml:183]: TODO: check migrations with dataset in old model or introduce breaking changes

File: deployment/factorycube-server/values.yaml [hunk 5]
```diff
@@ -84,7 +84,10 @@ grafanaproxy:
 ### mqtt-to-postresql ###
 mqtttopostgresql:
   enabled: true
-  image: unitedmanufacturinghub/mqtt-to-postgresql
+  image: 
+    repository: larswitte/mqtt-to-postgresql
```
Comment (JeremyTheocharis) [inline deployment/factorycube-server/values.yaml:88]: use unitedmanufacturinghub as repository

File: deployment/factorycube-server/values.yaml [hunk 6]
```diff
@@ -1537,7 +1540,7 @@ grafana:
   #     memory: 50Mi
 ### nodered ###
 nodered:
-  enabled: false 
+  enabled: true
```
Comment (JeremyTheocharis) [inline deployment/factorycube-server/values.yaml:1543]: leave it disabled by default

**Review by JeremyTheocharis — CHANGES_REQUESTED** [review]

**Comment by JeremyTheocharis** [comment]
Todo: adjust documentation

**Comment by larswitte** [comment]
***!!!Breaking change of postgres: documentation change necessary!!!***

**Review by JeremyTheocharis — APPROVED** [review]

**Pull request merged**